crates/bitwarden-core/src/auth/pin.rs

```diff
@@ -0,0 +1,101 @@
+use bitwarden_crypto::{EncString, PinKey};
+
+use crate::{
+    client::{LoginMethod, UserLoginMethod},
+    error::{Error, Result},
+    Client,
+};
+
+pub(crate) fn validate_pin(
+    client: &Client,
+    pin: String,
+    pin_protected_user_key: EncString,
+) -> Result<bool> {
+    let login_method = client
+        .internal
+        .get_login_method()
+        .ok_or(Error::NotAuthenticated)?;
+
+    #[allow(irrefutable_let_patterns)]
+    let LoginMethod::User(login_method) = login_method.as_ref() else {
+        return Err(Error::NotAuthenticated);
+    };
+
+    match login_method {
+        UserLoginMethod::Username { email, kdf, .. }
+        | UserLoginMethod::ApiKey { email, kdf, .. } => {
```

Comment:
**question:** Do we expect Api key logins to use pin?

Reply:
I don't think so, and I'm not sure if they would need password validation either. I think this is mostly a case of copying over the `match` to get `email` and `kdf`. We can return an error for `ApiKey` if you prefer though.